core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1536,29 +1537,38 @@ class ReplicaManager(val config: KafkaConfig,
     }
   }
 
-  def electPreferredLeaders(controller: KafkaController,
-                            partitions: Set[TopicPartition],
-                            responseCallback: Map[TopicPartition, ApiError] => Unit,
-                            requestTimeout: Long): Unit = {
+  def electLeaders(
```

Comment:
There doesn't seem to be a good reason for this to be part of `ReplicaManager` instead of `KafkaController`. No need to do it here, but maybe we can move the election purgatory out of `ReplicaManager` in a follow-up patch.

Reply:
https://issues.apache.org/jira/browse/KAFKA-8408